File: backend/app/services/admin_service.py
from fastapi import HTTPException, status

from app.config import dynamodb


users_table = dynamodb.Table("Users")
lost_items_table = dynamodb.Table("LostItems")
found_items_table = dynamodb.Table("FoundItems")


class AdminService:

    # ==========================================
    # USERS
    # ==========================================

    @staticmethod
    def get_all_users():

        response = users_table.scan()

        users = []

        for user in response.get("Items", []):

            users.append({
                "user_id": user.get("user_id"),
                "name": user.get("name"),
                "email": user.get("email"),
                "role": user.get("role", "user"),
                "flagged": user.get("flagged", False),
            })

        return users


    @staticmethod
    def get_user(user_id: str):

        response = users_table.get_item(
            Key={
                "user_id": user_id
            }
        )

        user = response.get("Item")

        if not user:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )

        return {
            "user_id": user.get("user_id"),
            "name": user.get("name"),
            "email": user.get("email"),
            "role": user.get("role", "user"),
            "flagged": user.get("flagged", False),
        }


    @staticmethod
    def delete_user(user_id: str):

        response = users_table.get_item(
            Key={
                "user_id": user_id
            }
        )

        if "Item" not in response:

            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )

        users_table.delete_item(
            Key={
                "user_id": user_id
            }
        )

        return {
            "message": "User deleted successfully"
        }


    @staticmethod
    def flag_user(
        user_id: str,
        flagged: bool
    ):

        response = users_table.get_item(
            Key={
                "user_id": user_id
            }
        )

        user = response.get("Item")

        if not user:

            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail="User not found",
            )

        users_table.update_item(
            Key={
                "user_id": user_id
            },
            UpdateExpression="SET flagged = :flagged",
            ExpressionAttributeValues={
                ":flagged": flagged
            },
            ReturnValues="ALL_NEW",
        )

        return {
            "message":
                "User flagged successfully"
                if flagged
                else
                "User unflagged successfully",

            "user_id": user_id,

            "flagged": flagged,
        }


    # ==========================================
    # ITEMS
    # ==========================================

    @staticmethod
    def get_all_items():

        lost_response = lost_items_table.scan()
        found_response = found_items_table.scan()

        lost_items = lost_response.get("Items", [])
        found_items = found_response.get("Items", [])

        items = []

        # -----------------------------
        # LOST ITEMS
        # -----------------------------

        for item in lost_items:

            items.append({
                "item_id": item.get("item_id"),
                "item_name": item.get("item_name"),
                "category": item.get("category"),
                "description": item.get("description"),
                "location": item.get("location"),
                "contact_name": item.get("contact_name"),
                "contact_phone": item.get("contact_phone"),
                "status": "Lost",
                "date_lost": item.get("date_lost"),
                "date_found": None,
            })


        # -----------------------------
        # FOUND ITEMS
        # -----------------------------

        for item in found_items:

            items.append({
                "item_id": item.get("item_id"),
                "item_name": item.get("item_name"),
                "category": item.get("category"),
                "description": item.get("description"),
                "location": item.get("location"),
                "contact_name": item.get("contact_name"),
                "contact_phone": item.get("contact_phone"),
                "status": "Found",
                "date_lost": None,
                "date_found": item.get("date_found"),
            })


        return items


    @staticmethod
    def delete_item(item_id: str):

        # Check LostItems first

        lost_response = lost_items_table.get_item(
            Key={
                "item_id": item_id
            }
        )

        if "Item" in lost_response:

            lost_items_table.delete_item(
                Key={
                    "item_id": item_id
                }
            )

            return {
                "message": "Lost item deleted successfully"
            }


        # Check FoundItems

        found_response = found_items_table.get_item(
            Key={
                "item_id": item_id
            }
        )

        if "Item" in found_response:

            found_items_table.delete_item(
                Key={
                    "item_id": item_id
                }
            )

            return {
                "message": "Found item deleted successfully"
            }


        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Item not found",
        )

File: frontend/src/services/adminService.ts
import axios from "axios";

const API_URL = "http://127.0.0.1:8000";

interface AdminUser {
    user_id: string;
    name: string;
    email: string;
    role: string;
    flagged: boolean;
}

interface AdminItem {
    item_id: string;
    item_name: string;
    category: string;
    description: string;
    location: string;
    contact_name: string;
    contact_number?: string;
    contact_phone?: string;
    status: "Lost" | "Found";
    date_lost?: string;
    date_found?: string;
}

function getAuthHeaders() {
    const token = localStorage.getItem("access_token");

    return {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
    };
}

/* =========================
   USERS
========================= */

export async function getAdminUsers(): Promise<AdminUser[]> {
    const response = await axios.get(
        `${API_URL}/admin/users`,
        {
            headers: getAuthHeaders(),
        }
    );

    return response.data?.data || [];
}

export async function getAdminUser(
    userId: string
): Promise<AdminUser> {
    const response = await axios.get(
        `${API_URL}/admin/users/${userId}`,
        {
            headers: getAuthHeaders(),
        }
    );

    return response.data?.data;
}

export async function flagAdminUser(
    userId: string,
    flagged: boolean
) {
    const response = await axios.put(
        `${API_URL}/admin/users/${userId}/flag`,
        null,
        {
            params: {
                flagged,
            },
            headers: getAuthHeaders(),
        }
    );

    return response.data;
}

export async function deleteAdminUser(
    userId: string
) {
    const response = await axios.delete(
        `${API_URL}/admin/users/${userId}`,
        {
            headers: getAuthHeaders(),
        }
    );

    return response.data;
}


/* =========================
   ITEMS
========================= */

export async function getAdminItems(): Promise<AdminItem[]> {
    const response = await axios.get(
        `${API_URL}/admin/items`,
        {
            headers: getAuthHeaders(),
        }
    );

    return response.data?.data || [];
}

export async function deleteAdminItem(
    itemId: string
) {
    const response = await axios.delete(
        `${API_URL}/admin/items/${itemId}`,
        {
            headers: getAuthHeaders(),
        }
    );

    return response.data;
}

export type {
    AdminItem, AdminUser
};

